query/hash_reduction.cu

```diff
@@ -0,0 +1,395 @@
+      map_type *const __restrict__ map,
+      uint8_t *dimInputValues,
+      size_t vectorCapacity,
+      uint8_t *dimOutputValues,
+      uint8_t dimWidthPrefixSum[NUM_DIM_WIDTH],
+      uint8_t *measureValues,
+      uint32_t *global_write_index) : map(map), dimInputValues(dimInputValues),
+                                      vectorCapacity(vectorCapacity),
+                                      dimOutputValues(dimOutputValues),
+                                      measureValues(measureValues),
+                                      global_write_index(
+                                          global_write_index) {
+    memcpy(this->dimWidthPrefixSum,
+           dimWidthPrefixSum, sizeof(uint8_t) * NUM_DIM_WIDTH);
+  }
+
+  __host__ __device__
+  ~ExtractGroupByResultFunctor() {
+  }
+
+  map_type *map;
+  uint8_t *dimInputValues;
+  uint32_t *global_write_index;
+  uint8_t *measureValues;
+  uint8_t *dimOutputValues;
+  size_t vectorCapacity;
+  uint8_t dimWidthPrefixSum[NUM_DIM_WIDTH];
+
+  // copy_dim_values moves the inputDimValues at inputIdx to curDimOutput
+  // at outputIdx.
+  __host__ __device__
+  void copy_dim_values(uint32_t inputIdx, size_t outputIdx) {
+    uint8_t * curDimInput = dimInputValues;
+    uint8_t * curDimOutput = dimOutputValues;
+
+    // idx in numDimsPerDimWidth;
+    int widthIdx = 0;
+    uint8_t
+    totalDims = dimWidthPrefixSum[NUM_DIM_WIDTH - 1];
+    // write values
+    // pointer address for dim value d on row r is
+    // (base_ptr + accumulatedValueBytes * vectorCapacity) + r * dimValueBytes.
+    for (uint8_t dimIndex = 0; dimIndex < totalDims; dimIndex += 1) {
+      // find correct widthIdx.
+      while (widthIdx < NUM_DIM_WIDTH
+          && dimIndex >= dimWidthPrefixSum[widthIdx]) {
+        widthIdx += 1;
+      }
+
+      uint16_t dimValueBytes = 1 << (NUM_DIM_WIDTH - widthIdx - 1);
+      curDimInput += dimValueBytes * inputIdx;
+      curDimOutput += dimValueBytes * outputIdx;
+      setDimValue(curDimOutput, curDimInput, dimValueBytes);
+
+      // set to the start of next dim value
+      curDimInput += (vectorCapacity - inputIdx) * dimValueBytes;
+      curDimOutput += (vectorCapacity - outputIdx) * dimValueBytes;
+    }
+
+    // write nulls.
+    // Now both curDimInput and curDimOutput should point to start of
+    // null vector.
+    for (uint8_t dimIndex = 0; dimIndex < totalDims; dimIndex += 1) {
+      curDimInput += inputIdx;
+      curDimOutput += outputIdx;
+
+      *curDimOutput = *curDimInput;
+
+      // set to the start of next dim null vector
+      curDimInput += vectorCapacity - inputIdx;
+      curDimOutput += vectorCapacity - outputIdx;
+    }
+  }
+
+  __host__ __device__
+  void operator()(const int i) {
+    key_type unusedKey = map->get_unused_key();
+    key_value_ptr_type mapValues = map->data();
+    size_t mapSize = map->capacity();
+    if (i < mapSize) {
+      key_type current_key = mapValues[i].first;
+      if (current_key != unusedKey) {
```

Comment:
is it better to use `key_equal`?

Reply:
no, key_equal here is compare the hash part (higher 32 bit) of the key, here we need to compare to whole key

Reply:
i see.